tripy/examples/diffusion/example.py

```diff
@@ -0,0 +1,222 @@
+        tp.InputInfo(unconditional_context_shape, dtype=tp.float32),
+        tp.InputInfo(conditional_context_shape, dtype=tp.float32),
+        tp.InputInfo(latent_shape, dtype=tp.float32),
+        tp.InputInfo((1,), dtype=tp.float32),
+        tp.InputInfo((1,), dtype=tp.float32),
+        tp.InputInfo((1,), dtype=tp.float32),
+        tp.InputInfo((1,), dtype=tp.float32),
+    )
+    unet_compile_end_time = time.perf_counter()
+    print(f"Compilation of UNet took {unet_compile_end_time - unet_compile_start_time} seconds.")
+
+    timesteps = list(range(1, 1000, 1000 // args.steps))
+    print(f"Running for {timesteps} timesteps.")
+    alphas = model.alphas_cumprod[tp.Tensor(timesteps)]
+    alphas_prev = tp.concatenate([tp.Tensor([1.0]), alphas[:-1]], dim=0)
+
+    # start with random noise
+    if args.seed is not None:
+        torch.manual_seed(args.seed)
+    torch_latent = torch.randn((1, 4, 64, 64)).to("cuda")
+    latent = tp.Tensor(torch_latent)
+
+    def run(model, unconditional_context, context, latent, timestep, alphas, alphas_prev, guidance):
+        return model(unconditional_context, context, latent, timestep, alphas, alphas_prev, guidance)
+
+    # This is diffusion
+    print("Running diffusion...")
+    unet_run_start = time.perf_counter()
+    for index, timestep in (t := tqdm(list(enumerate(timesteps))[::-1])):
+        t.set_description("idx: %1d, timestep: %3d" % (index, timestep))
+        tid = tp.Tensor([index])
+        latent = run(
+            compiled_model,
+            unconditional_context,
+            context,
+            latent,
+            tp.cast(tp.Tensor([timestep]), tp.float32),
+            alphas[tid],
+            alphas_prev[tid],
+            tp.Tensor([args.guidance]),
+        )
+    unet_run_end = time.perf_counter()
+    print(f"Finished running diffusion. Inference took {unet_run_end - unet_run_start} seconds.")
+
+    # Upsample latent space to image with autoencoder
+    print("Compiling VAE decoder...")
+    vae_compile_start_time = time.perf_counter()
+    vae_compiler = tp.Compiler(model.decode)
+    vae_decode = vae_compiler.compile(tp.InputInfo((1, 4, 64, 64), dtype=tp.float32))
+    vae_compile_end_time = time.perf_counter()
+    print(f"Compilation took {vae_compile_end_time - vae_compile_start_time} seconds.")
+
+    print(f"Decoding latent...")
+    vae_run_start = time.perf_counter()
+    x = vae_decode(latent)
+    # x = model.decode(latent)
+    vae_run_end = time.perf_counter()
+    print(f"Finished decoding latent. Inference took {vae_run_end - vae_run_start} seconds.")
+
+    run_end_time = time.perf_counter()
+    x.eval()
+    print(f"Full pipeline took {run_end_time - run_start_time} seconds.")
+
+    # save image
+    im = Image.fromarray(cp.from_dlpack(x).get().astype(np.uint8, copy=False))
+    print(f"saving {args.out}")
+    im.save(args.out)
+    # Open image.
+    if not args.noshow:
+        im.show()
+
+def hf_diffusion(args):
```

Comment:
Why do we want this?

Reply:
This allows a pure torch inference mode for comparison with the example script - I also plan to use it for the e2e accuracy test that compares the two output images.